### PR TITLE
Log password requirement details in demo environment

### LINKED ISSUE #4048: [FEATURE] Log more information about Password strength feature
**Is your feature request related to a problem?**
It was brought up that we log out that a password is weak, with no steps as to how to determine the strength of a password. Maybe linking to this: https://lowe.github.io/tryzxcvbn/, or at least logging out that we are using zxcvbn could help.

### PR DESCRIPTION
**Pull request by camerondurham**
### Description

Logs details about the password validations done when demo clusters are created. The current message "Password $ADMIN_PASSWORD is weak. Please re-try with a stronger password." doesn't provide enough information communicate what the user needs to set the password to.

* Category: Enhancement
* Why these changes are required?
   * See #4048 - to improve UX when creating test OpenSearch clusters by logging details on password requirements
* What is the old behavior before changes and new behavior after changes?
    * Before: OS 2.12 fails to start with message `Password sample@1029 is weak. Please re-try with a stronger password.`
    * After: security plugin fails to start with message `Password weakpassword failed validation with error "Weak password". Please re-try with a minimum 8 character password that is "strong" per zxcvbn validation.`

### Issues Resolved

Potentially can resolve #4048, though would need issue author @derek-ho to confirm this is the intended change.

### Testing

Updated existing unit test that tests "password validation failure". 

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented <-- In a separate PR, I also want to update relevant docs, including [docker getting started docs](https://opensearch.org/docs/latest/install-and-configure/install-opensearch/docker/#run-opensearch-in-a-docker-container) docs with perhaps the zxcvbn "validation" tool.
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java
Comment (camerondurham) [inline src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java:133]: Should I include more tests for each of the other validation failures? (e.g. too short password, weak password)

File: src/main/java/org/opensearch/security/support/ConfigConstants.java
Comment (camerondurham) [inline src/main/java/org/opensearch/security/support/ConfigConstants.java:276]: I might be using "DEFAULT" incorrectly here, happy to move this back if this isn't the intended use of ConfigConstants.

**Review by peternied — COMMENTED** [review]
Thanks for the pull request @camerondurham - minor comments

File: src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java:133]: You can add 1 more test around password length validation, otherwise the test-cases should suffice.

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (camerondurham) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:152]: Thanks, will do. I realized these requirements are a bit more clearly spelled out in [these docs](https://opensearch.org/docs/latest/security/configuration/yaml/#password-settings) that I missed, so I'm add something like this:

> Password %s failed validation: "%s". Please re-try with a minimum %d character password and must contain at least one uppercase letter, one lowercase letter, one digit, and one special character that is strong. Password strength can be tested here: https://lowe.github.io/tryzxcvbn

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (camerondurham) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:135]: Taking a closer look at this, the [docs](https://opensearch.org/docs/latest/security/configuration/yaml/#password-settings) say this is configurable via `plugins.security.restapi.password_min_length` but is it actually configurable if it's being set to `8` all the time? I might be totally missing something though.

File: src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java
Comment (camerondurham) [reply src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java]: Thanks! Added a "too short" test case in revision 2.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java
Comment (camerondurham) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: I see, good point. I've removed `ConfigConstants.java` from the change set and moved things back to where they're used.

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (camerondurham) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java]: Godd point. I've added the test url and described the validation regex in rev 2.

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (peternied) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:135]: It would be strange to have different requirements in this tool from the OpenSearch instance that is being setup.

@DarshitChanpura Can you create an issue around this?  The settings configuration should be read from a singular OpenSearch settings, not a separate setting configuration or hardcoded.

Comment (derek-ho) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:135]: +1, not sure but it seems like the regex also has some concept of length, maybe we can combine them?

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (derek-ho) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:151]: I believe that the minimum strength can also be configured with `plugins.security.restapi.password_score_based_validation_strength`, so not sure if we want to hard code strong here? Also not sure how I feel about adding hard coded links into a log message, but maybe its fine, will defer to others here.

Comment (derek-ho) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:151]: More context is the feature was added in this PR: https://github.com/opensearch-project/security/issues/2569 - I pulled this setting from the PR description

Comment (derek-ho) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:151]: Also pretty sure the regex can be changed to not be one uppercase letter, one digit, etc. maybe we should just leave it at "meeting the minimum regex and password strength requirements"? But not sure since I think there will be a decent chunk of people who are using the minimum.

**Comment by derek-ho** [comment]
Thanks for the contribution @camerondurham !

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:135]: `plugins.security.restapi.password_min_length` was introduced in https://github.com/opensearch-project/security/pull/2557 for password validation. 

The password length is not configurable for admin user since the demo script doesn't read settings from opensearch.yml as it is executed before reading the yml file. Remember, admin password is changeable, and the password validation introduced in #2557 should come into play when verifying new password. For changes related to demo script, specifying that the password has to be 8 characters or longer should be enough. 

Let me know if I missed something here.

File: src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java
Comment (DarshitChanpura) [inline src/test/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurerTests.java:155]: Create a class-level variable:
```java
String PASSWORD_VERIFICATION_FAILURE_MSG = "Password %s failed validation: \"%s\". Please re-try with a minimum %d character password and must contain at least one uppercase letter, one lowercase letter, one digit, and one special character that is strong. Password strength can be tested here: https://lowe.github.io/tryzxcvbn";
```

and then can be used here (and above as):

```java
verifyStdOutContainsString(
            String.format(
                PASSWORD_VERIFICATION_FAILURE_MSG,
                "short",
                INVALID_PASSWORD_TOO_SHORT.message(),
                DEFAULT_PASSWORD_MIN_LENGTH
            )
```

**Review by DarshitChanpura — COMMENTED** [review]
@camerondurham the settings `plugins.security.restapi.password_min_length` and `plugins.security.restapi.password_score_based_validation_strength` come into play once the security plugin is setup and cluster is ready, which is when all the config settings are read.

However, in case of demo configuration script execution, these settings are not available to be read, as the tools write opensearch.yml with some default settings. 

In case if we try to pass these settings in opensearch.yml, the demo script will stop execution since it thinks that security plugin is already setup because it is able to find `plugins.security.<setting>` already defined in opensearch.yml ([check this out](https://github.com/opensearch-project/security/blob/26b224fa1e2986431f22cf865592c4ce6fa6f49c/src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java#L109))

I think we are overthinking this. The regex + password length supplied here are one time use before a fresh cluster start-up. We can leave those settings as is and modify the log message to focus on the scope of this PR. Thoughts?

**Review by DarshitChanpura — COMMENTED** [review]
Thank you for addressing all the feedback! Nothing more from my end; waiting for CI to be green before approval.

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (derek-ho) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:151]: Yea got your point. I had a misunderstanding that this was being applied to all user creation/password validation, in which case those custom settings may be present. @DarshitChanpura you are right that that would not be the case, or else they would never make it into this demo script tool. Sorry for the misleading comment.

Comment (camerondurham) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:151]: No worries at all, I had the same misunderstanding. Your comment wasn't misleading :)

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (peternied) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:135]: >The password length is not configurable for admin user since the demo script doesn't read settings from opensearch.yml as it is executed before reading the yml file. Remember, admin password is changeable, and the password validation introduced in #2557 should come into play when verifying new password. For changes related to demo script, specifying that the password has to be 8 characters or longer should be enough.

I'm not saying we need to implement it, but we should document that the password requirements for this tool that ships with OpenSearch does not use the same settings.  If someone was to implement this functionality I would approve merging it, what do you think @DarshitChanpura ?

**Review by peternied — APPROVED** [review]

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:135]: done. https://github.com/opensearch-project/security/issues/4081

**Review by DarshitChanpura — APPROVED** [review]

**Comment by DarshitChanpura** [comment]
Thank you for this contribution @camerondurham !

**Pull request merged**
